pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,410 @@
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		types.PutSeriesMetadataSlice(innerSeries)
+	}()
+
+	if len(innerSeries) == 0 {
+		// No input series == no output series.
+		return nil, nil
+	}
+
+	h.innerSeriesMetricNames.CaptureMetricNames(innerSeries)
+	groups := map[string]groupWithLabels{}
+	h.seriesGroupPairs = make([]seriesGroupPair, len(innerSeries))
+	if err != nil {
+		return nil, err
+	}
+	b := make([]byte, 0, 1024)
+	lb := labels.NewBuilder(labels.EmptyLabels())
+
+	for innerIdx, series := range innerSeries {
+		// Each series belongs to two groups, one without the `le` label, and one with all labels.
+		// Sometimes these are the same group.
+
+		// Store the le label. If it doesn't exist, it'll be an empty string
+		le := series.Labels.Get(labels.BucketLabel)
+		h.seriesGroupPairs[innerIdx].bucketValue = le
+
+		// First get the group with all labels
+		b = series.Labels.Bytes(b)
+		g, groupExists := groups[string(b)]
+		if !groupExists {
+			g.labels = series.Labels
+			g.group = bucketGroupPool.Get()
+			groups[string(b)] = g
+		}
+		g.group.lastInputSeriesIdx = innerIdx
+		g.group.remainingSeriesCount++
+		h.seriesGroupPairs[innerIdx].nativeHistogramGroup = g.group
+
+		// Then get the group without the `le` label. This may be the same
+		// as the previous group if not le label exists.
+		// We still need to do this (rather than set it to nil) so that we know when to emit
+		// NewBadBucketLabelWarning when the series are processed.
+		b = series.Labels.BytesWithoutLabels(b, labels.BucketLabel)
+		g, groupExists = groups[string(b)]
+
+		if !groupExists {
+			lb.Reset(series.Labels)
+			lb.Del(labels.BucketLabel)
+			g.labels = lb.Labels()
+			g.group = bucketGroupPool.Get()
+			groups[string(b)] = g
+		}
+		g.group.lastInputSeriesIdx = innerIdx
+		g.group.remainingSeriesCount++
+		h.seriesGroupPairs[innerIdx].classicHistogramGroup = g.group
+	}
+
+	seriesMetadata := types.GetSeriesMetadataSlice(len(groups))
+	h.remainingGroups = make([]*bucketGroup, 0, len(groups))
+	for _, g := range groups {
+		seriesMetadata = append(seriesMetadata, types.SeriesMetadata{Labels: g.labels.DropMetricName()})
+		h.remainingGroups = append(h.remainingGroups, g.group)
+	}
+
+	// Sort the groups by the last series within them. This helps finish a group
+	// as soon as possible when accumulating them.
+	sort.Sort(bucketGroupSorter{seriesMetadata, h.remainingGroups})
+
+	return seriesMetadata, nil
+}
+
+func (h *HistogramFunctionOverInstantVector) NextSeries(ctx context.Context) (types.InstantVectorSeriesData, error) {
+	if len(h.remainingGroups) == 0 {
+		// No more groups left.
+		return types.InstantVectorSeriesData{}, types.EOS
+	}
+
+	// Determine next group to return
+	thisGroup := h.remainingGroups[0]
+	h.remainingGroups = h.remainingGroups[1:]
+	defer func() {
+		// Reset the group before returning to the pool
+		//thisGroup.groupedMetricName = ""
+		thisGroup.lastInputSeriesIdx = 0
+		if thisGroup.pointBuckets != nil {
```

Comment:
There's no need for this check here - `LimitingBucketedPool.Put` already handles the case where the slice is `nil`.